### ISSUE TITLE
[Regression] preferKeyPath option introduces compile time issue

### ISSUE THREAD

**Issue by Panajev** [nicklockwood/SwiftFormat#694]
Hello @nicklockwood ,

We have noticed the following transformation as result of the preferKeyPath option being activated:

<img width="1213" alt="Screenshot 2020-08-05 at 23 12 52" src="https://user-images.githubusercontent.com/562991/89469374-32b1a280-d771-11ea-92ed-564f028e8271.png">

where the compiler error presented is: **Key path value type 'Game' cannot be converted to contextual type 'Game?'**

Other replacements worked fine, but a couple of them gave us this error and for now we are disabling the rule. Is this a known issue?

Kind Regards,

Goffredo

**Comment by nicklockwood** [comment]
@Panajev yes, this is a known issue (I just added it to the README). The simple workaround in these cases is to replace `compactMap` with `map`, since there is no need to use `compactMap` in cases where the return value isn't Optional.

**Comment by Panajev** [comment]
Shall I close this or is it something that you think can be fixed at some point and this is best kept open?

**Comment by nicklockwood** [comment]
@Panajev I'll leave it open so others can find it. I'll try to gauge how common the problem is, and I may consider adding an option to disable the feature just for `compactMap()`.

I don't see any simple way for me to resolve it, but I'm hoping it's something Apple will eventually fix in Swift itself.

**Comment by rogerluan** [comment]
+1 

My use case is when the return type is optional, e.g.:

```swift
// Before
colors.compactMap { $0.cgColor }
// After
colors.compactMap(\.cgColor) // Error: Key path value type 'CGColor' cannot be converted to contextual type 'CGColor?'

// Before
urls.compactMap { $0.lastPathComponent }
// After
urls.compactMap(\.lastPathComponent) // Error: Key path value type 'String' cannot be converted to contextual type 'String?'
```

Perhaps the option to be on the safe side and disable the feature just for `compactMap` isn't too bad 🤷‍♂️ The workaround for us is to either sprinkle `// swiftformat:disable:next preferKeyPath` throughout the codebase, or disable the rule completely

**Comment by rogerluan** [comment]
You're totally right @nicklockwood , I apologize, as I thought the properties were optional 😮 but they're not, in both cases. So changing from compactMap to map worked (and, I must add that IMO the previous use of compactMap was wrong). Thanks for pointing it out! 🎉

**Comment by OliverPearmain-Nutmeg** [comment]
FYI something like the below can also cause problems...

```
public protocol Model {
    var text: String { get }
}

enum Example {
    func modelPublisher() -> AnyPublisher<Model, Error> {

    }
    func textPublisher() -> AnyPublisher<String?, Error> {
        modelPublisher()
            .map(\.text) // ERROR: Key path value type 'String' cannot be converted to contextual type 'String?'
            .eraseToAnyPublisher()
    }
}
```

This can be fixed by adjusting the signature of `textPublisher` to return `AnyPublisher<String, Error>` but sometimes it might not be possible to do that for example if the method signature is defined in a `protocol` and other concrete instances have a valid reason to emit a nil.
